srv/src/main/java/my/bookshop/handlers/HierarchyHandler.java

```diff
@@ -0,0 +1,236 @@
+        Integer id = (Integer) r.get(key);
+        if (id != null) {
+            ancestorIds.add(id);
+        }
+    } 
+
+    private List<GenreHierarchy> topLevels(CqnTopLevelsTransformation topLevels, CqnPredicate filter) {
+        return topLevels.levels() < 0 ? topLevelsAll(filter) : topLevelsLimit(topLevels.levels(), filter);
+    }
+
+    private List<GenreHierarchy> topLevelsLimit(long limit, CqnPredicate filter) {
+        Map <Integer, GenreHierarchy> lookup = new HashMap<>();
+
+        CqnSelect getRoots = Select.from(GENRE_HIERARCHY).where(gh -> gh.parent_id().eq(0).and(filter));
+        List<GenreHierarchy> roots = db.run(getRoots).listOf(GenreHierarchy.class);
+        roots.forEach(root -> {
+            root.setDistanceFromRoot(0l);
+            lookup.put(root.getNodeId(), root);
+            List<Integer> parents = List.of(root.getNodeId());
+            for (long i = 1; i < limit; i++) {
+                List<Integer> ps = parents;
+                CqnSelect getChildren = Select.from(GENRE_HIERARCHY).where(gh -> gh.parent_id().in(ps).and(filter));
+                List<GenreHierarchy> children = db.run(getChildren).listOf(GenreHierarchy.class);
+                if (children.isEmpty()) {
+                    break;
+                }
+                long dfr = i;
+                parents = children.stream().peek(gh -> {
+                    gh.setParent(lookup.get(gh.getParentId()));
+                    gh.setDistanceFromRoot(dfr);
+                    lookup.put(gh.getNodeId(), gh);
+                }).map(GenreHierarchy::getNodeId).toList();
+            }
+        });
+
+        return lookup.values().stream().sorted(new Sorter()).toList();
+    }
+
+    private List<GenreHierarchy> topLevelsAll(CqnPredicate filter) {
+        Map<Integer, GenreHierarchy> lookup = new HashMap<>();
+
+        CqnSelect allCqn = Select.from(GENRE_HIERARCHY).where(filter);
+        var all = db.run(allCqn).listOf(GenreHierarchy.class);
+        all.forEach(gh -> lookup.put(gh.getNodeId(), gh));
+        all.forEach(gh -> gh.setParent(lookup.get(gh.getParentId())));
+        all.forEach(gh -> gh.setDistanceFromRoot(dfr(gh)));
+
+        return all.stream().sorted(new Sorter()).toList();
+    }
+
+    private static long dfr(GenreHierarchy gh) {
+        long dfr = 0;
+        while (gh.getParent() != null) {
+            dfr++;
+            gh = gh.getParent();
+        }
+
+        return dfr;
+    }
+
+    private class Sorter implements Comparator<GenreHierarchy> {
+
+        @Override
+        public int compare(GenreHierarchy o1, GenreHierarchy o2) {
+            return toPath(o1).compareTo(toPath(o2));
+        }
+
+        String toPath(GenreHierarchy gh) {
+            String path;
+
+            path = gh.getName();
+            while (gh.getParent() != null) {
+                gh = gh.getParent();
+                path = gh.getName() + "." + path;
+            }
+
+            return path;
```

Comment:
I think this (i.e. my ☺️) algorithm is buggy, unfortunately. We need to compare level by level